sdk/servicebus/azure-servicebus/azure/servicebus/_pyamqp/aio/_receiver_async.py

```diff
@@ -103,18 +141,50 @@ async def _outgoing_disposition(
         settled: Optional[bool],
         state: Optional[Union[Received, Accepted, Rejected, Released, Modified]],
         batchable: Optional[bool],
+        *,
+        on_disposition: Optional[Callable] = None,
     ):
         disposition_frame = DispositionFrame(
             role=self.role, first=first, last=last, settled=settled, state=state, batchable=batchable
         )
         if self.network_trace:
             _LOGGER.debug("-> %r", DispositionFrame(*disposition_frame), extra=self.network_trace_params)
+
+        # If trying to settle a message, keep track of the disposition
+        if on_disposition:
+            delivery = PendingDisposition(
+                on_delivery_settled = on_disposition,
+                frame=disposition_frame,
+                settled=settled,
+                transfer_state=state,
+                start=time.time(),
+                sent=True,
+            )
+            self._pending_receipts.append(delivery)
+
         await self._session._outgoing_disposition(disposition_frame) # pylint: disable=protected-access
 
+    async def _incoming_disposition(self, frame):
+        # If delivery_id is not settled, return
+        # if not frame[3]:  # settled
+        #     return
+        range_end = (frame[2] or frame[1]) + 1  # first or last
+        settled_ids = list(range(frame[1], range_end))
+        unsettled = []
+        for delivery in self._pending_receipts:
+            if delivery.sent and delivery.frame[1] in settled_ids:
+                await delivery.on_settled(LinkDeliverySettleReason.DISPOSITION_RECEIVED, frame[4])  # state
+                continue
+            unsettled.append(delivery)
+        self._pending_receipts = unsettled
```

Comment:
We may need to look at the concurrency implications here.

Reply:
note for future: We do a similar format in incoming_disp on sender so will need to consider that there too